File: tools/ghidra_scripts/FindCDXrefs.java
// @author
// @category Search
// @keybinding
// @menupath Search.Find CD Xrefs
// @toolbar

import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Instruction;
import ghidra.program.model.symbol.Reference;
import ghidra.program.model.symbol.ReferenceIterator;

public class FindCDXrefs extends GhidraScript {

    private void dumpXrefs(String label, long va) throws Exception {
        Address addr = currentProgram.getAddressFactory().getDefaultAddressSpace().getAddress(va);
        println("\n\n===== " + label + " @ " + addr + " =====");

        ReferenceIterator refs = currentProgram.getReferenceManager().getReferencesTo(addr);
        int count = 0;
        while (refs.hasNext()) {
            Reference ref = refs.next();
            Address from = ref.getFromAddress();
            count++;
            println("-- xref #" + count + " from " + from + " (" + ref.getReferenceType() + ") --");

            var func = currentProgram.getFunctionManager().getFunctionContaining(from);
            if (func != null) {
                println("   in function: " + func.getName() + " @ " + func.getEntryPoint());
            }

            Instruction instr = currentProgram.getListing().getInstructionAt(from);
            if (instr != null) {
                println("   Context before:");
                Instruction ctx = instr;
                for (int i = 0; i < 8; i++) {
                    ctx = currentProgram.getListing().getInstructionBefore(ctx.getAddress());
                    if (ctx == null) break;
                }
                // walk forward from earliest to print in order
                java.util.List<String> lines = new java.util.ArrayList<>();
                Instruction walk = ctx != null ? ctx : instr;
                for (int i = 0; i < 20 && walk != null; i++) {
                    String marker = walk.getAddress().equals(from) ? " <=== XREF" : "";
                    lines.add("     " + walk.getAddress() + ": " + walk.toString() + marker);
                    walk = currentProgram.getListing().getInstructionAfter(walk.getAddress());
                    if (walk != null && walk.getAddress().subtract(from) > 40) break;
                }
                for (String l : lines) println(l);
            }
        }
        if (count == 0) {
            println("   (no code xrefs found)");
        }
    }

    @Override
    public void run() throws Exception {
        long base = currentProgram.getImageBase().getOffset();
        println("Image base: 0x" + Long.toHexString(base));

        // VAs computed from PE section headers (CODE rva=0x1000/raw=0x600, DATA rva=0x6C000/raw=0x6B000)
        dumpXrefs("CD message copy 1 (file 0x6B4BD)", base + 0x46C4BDL - 0x400000L);
        dumpXrefs("CD message copy 2 (file 0x70430)", base + 0x471430L - 0x400000L);
        dumpXrefs("-eggplant copy 1 (file 0x6B22B)", base + 0x46C22BL - 0x400000L);
        dumpXrefs("-eggplant copy 2 (file 0x7020A)", base + 0x47120AL - 0x400000L);
        dumpXrefs("dummy (file 0x70204)", base + 0x471204L - 0x400000L);
    }
}

File: tools/ghidra_scripts/find_cd_xrefs.py
# Find xrefs to CD-message / eggplant / dummy strings
# @category Search

from ghidra.program.model.address import Address

def dump_xrefs(label, va):
    addr = currentProgram.getAddressFactory().getDefaultAddressSpace().getAddress(va)
    print("\n\n===== %s @ %s =====" % (label, addr))
    refs = currentProgram.getReferenceManager().getReferencesTo(addr)
    count = 0
    for ref in refs:
        count += 1
        frm = ref.getFromAddress()
        print("-- xref #%d from %s (%s) --" % (count, frm, ref.getReferenceType()))
        func = currentProgram.getFunctionManager().getFunctionContaining(frm)
        if func is not None:
            print("   in function: %s @ %s" % (func.getName(), func.getEntryPoint()))
        instr = currentProgram.getListing().getInstructionAt(frm)
        if instr is not None:
            ctx = instr
            for i in range(8):
                prev = currentProgram.getListing().getInstructionBefore(ctx)
                if prev is None:
                    break
                ctx = prev
            walk = ctx
            for i in range(20):
                if walk is None:
                    break
                marker = " <=== XREF" if walk.getAddress().equals(frm) else ""
                print("     %s: %s%s" % (walk.getAddress(), walk.toString(), marker))
                nxt = currentProgram.getListing().getInstructionAfter(walk)
                walk = nxt
    if count == 0:
        print("   (no code xrefs found)")

base = currentProgram.getImageBase().getOffset()
print("Image base: 0x%x" % base)

dump_xrefs("CD message copy 1 (file 0x6B4BD)", base + 0x6C4BD)
dump_xrefs("CD message copy 2 (file 0x70430)", base + 0x71430)
dump_xrefs("-eggplant copy 1 (file 0x6B22B)", base + 0x6C22B)
dump_xrefs("-eggplant copy 2 (file 0x7020A)", base + 0x7120A)
dump_xrefs("dummy (file 0x70204)", base + 0x71204)
